### PR TITLE
Image block: Move UI for lightbox from sidebar to the content toolbar alongside link settings

### PR DESCRIPTION
**Pull request by artemiomorales**
## What?

This PR moves the UI for the image lightbox from the sidebar's block settings to the context toolbar so that it can exist alongside the rest of the link logic.

## Why?

Addresses https://github.com/WordPress/gutenberg/issues/54916
Right now the UX for the lightbox is divorced from the rest of the handling for links.
This consolidates it and makes it more intuitive.

## How?

In the `image-url-input-ui` component, this PR adds the logic, markup, and styles for the new design and `Expand on click` toolbar item. Also, in `image.js`, I modified the logic for reading from or overriding the lightbox setting from the global styles.

#### With global styles `expand on click` setting _enabled_
1. In the global styles for the image block, ensure that `Expand on click` is enabled
2. Add an image to a post
3. Click the image and view the link settings; see that the `Expand on click` option is active
4. Publish and view the post; ensure the lightbox works as expected
6. In the post editor, click the image, open the link settings, and remove `Expand on click`
7. Check the post code source and see that `"lightbox":{"enabled":false}` has been added in the block delimiter
8. Save and view the post; ensure the lightbox is now disabled
9. In the post editor, reenable `Expand on click`
10. Check the code source again and verify `"lightbox":{"enabled":false}` has been removed

#### With global styles `expand on click` setting _disabled_
(same as above, but in reverse)
1. In the global styles for the image block, ensure that `Expand on click` is disabled
2. Add an image to a post
3. Click the image and view the link settings; see that the `Expand on click` option is deactivated
4. Publish and view the post; ensure the lightbox does not open
6. In the post editor, click the image, open the link settings, and enable `Expand on click`
7. Check the post code source and see that `"lightbox":{"enabled":true}` has been added in the block delimiter
8. Save and view the post; ensure the lightbox is now enabled
9. In the post editor, redeactivate `Expand on click`
10. Check the code source again and verify that the `"lightbox":{"enabled":true}` has been removed

### Site editor

For the site editor, please test the same as above. In addition, verify that toggling the global styles `Expand on ciick` setting also toggles the link UI for the image.

toggle-lightbox-global-styles

_Also_, check that adding and removing `Link to image file`, `Link to attachment page`, and custom links still works as expected.

### Testing Instructions for Keyboard

This uses existing UI and should work as expected with standard keyboard navigation.

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/5360536/4f408215-c156-4547-a22f-da1e7fd54fdb

### CODE REVIEW

**Comment by jasmussen** [comment]
A perhaps premature review, here's a GIF showing the move of expand on click to the toolbar:

state

Looks great, works great, and will pair well with lightbox working best when it's _on by default as set by the them_ and something you _untoggle on a per image basis_. There are some UX issues with the link UI toolbar, I suspect most of them are unrelated to this PR (dark border, the height, unifying the UI for linking and unlinking across the 3 types of link tokens, etc). Those shouldn't block this PR, but it'd be nice if some of the separators connected properly, and icons were aligned across. CC: @getdave and @richtabor  for awareness.

**Comment by gziolo** [comment]
This new flow looks much more intuitive. Would it make sense to put an expand icon marker on the image to replicate better how it looks on the front end and to emphasize more this selection in the editor?

**Comment by jasmussen** [comment]
> Would it make sense to put an expand icon marker on the image to replicate better how it looks on the front end and to emphasize more this selection in the editor?

Can make sense, though I would not make it a priority, and it can be entirely separate.

**Comment by getdave** [comment]
@richtabor and I are working through Link UI issues. You can follow along at https://github.com/WordPress/gutenberg/issues/50891

**Comment by jasmussen** [comment]
@artemiomorales @gziolo just checking in here, I'm super anxious and excited to get this PR in, and I'm eyeing a backport date of February 13th. Let me or anyone from @WordPress/gutenberg-design know if we can help with this one, it's quite important to get this flow right, so thanks so much for working on it 🙏

**Comment by getdave** [comment]
This is on my review list. If there's any UX/Design to review here then perhaps @richtabor would be well placed given he and I have been working on the Link UI together.

**Comment by richtabor** [comment]
I pushed some changes to get it closer. @artemiomorales, two follow-ups we need: 

1. Make the chevronDown toggle not render when lightbox is enabled
2. Style the selected state of the lightbox to emulate the updated styling I pushed. 

CleanShot 2024-01-19 at 15 01 20

Ideally this should eventually leverage LinkControl, so there can be proper suggestions, and linking within WordPress is the same everywhere. I imagine these controls much like "Create page" is, additional at the foot of the suggestions: 

CleanShot 2024-01-19 at 13 29 41

**Comment by artemiomorales** [comment]
> I pushed some changes to get it closer. @artemiomorales, two follow-ups we need:
>
> 1. Make the chevronDown toggle not render when lightbox is enabled
> 2. Style the selected state of the lightbox to emulate the updated styling I pushed.

@richtabor Great, thank you 🙏
Will take a look at those remaining issues

**Comment by artemiomorales** [comment]
> Ideally this should eventually leverage LinkControl, so there can be proper suggestions, and linking within WordPress is the same everywhere. I imagine these controls much like "Create page" is, additional at the foot of the suggestions:

@richtabor This is something to handle in a different PR, right?

**Review by artemiomorales — COMMENTED** [review]
I've set this PR as ready for review. Please check to see if how I've added the styles and markup for the `Expand on click` state of the popover makes sense. Thanks! 🙏

**Review by jasmussen — APPROVED** [review]
This is a massive step forward. Thank you. Here's a quick GIF showing the control when lightbox is off by default in theme.json, with all of the controls move to the Link button:

lightbox

Noting also that this works correctly when lightbox is enabled by default, it shows the link control linked to expand by default:
linked by default

Simply based on how much more intuitive this behavior is, this PR is ready to land.

---

Rich gives some good suggestions for followups in [this comment](https://github.com/WordPress/gutenberg/pull/57608#issuecomment-1901041892), I would not say those are blocking, and they can certainly be refined in the beta period.

To expand a bit, these are the main pieces of UI to polish, and I'll defer to Rich and Dave on how to best do that since it involves the link UI:
popover link to image file

popover link to attachment page

popopover expand on click

popover link dialog

Some notes on those, most notably that the link dialog should sit 12px below the block toolbar, like other popovers:
more options popover

Ideally we have helper text for image file and attachment options as well, so the three options are equally weighted.

And finally, if we can fidget the line-height of the _linked state_ so the overall popover is exactly 48px tall to match the toolbar, that would be great. That's this one:

popopover expand on click

In general the inner padding, vertical alignment of icons and controls inside can be polished a bit to be balanced all around. 

But as noted, not a blocker, and ideally some work I can help with. Thank you!

**Comment by gziolo** [comment]
Two notes on the technical side. Unit tests fail on CI due to changes in WP core that require a follow-up bug fix. There are some e2e test failures reported, too. It looks like some UI labels have changed in this PR so that also needs to be reflected in the tests.

**Comment by artemiomorales** [comment]
Just merged this 👍
Now that the lightbox UI is stable, we should reenable and update the tests for the lightbox. Will do that in a follow up PR.

**Comment by jasmussen** [comment]
Appreciate the push, thank you! What backporting labels need to be present here for it to land in 6.5?

**Comment by gziolo** [comment]
@c4rl0sbr4v0, can we still include it in the plugin release to test on production before syncing to WP core?

**Comment by jeryj** [comment]
👋🏻 The toolbar link icon still says "Add Link" when the Expand on Click option is enabled. Could the Add Link/Edit Link state of the toolbar link icon be synced up to the Expand on Click option state?

**Comment by cbravobernal** [comment]
> @c4rl0sbr4v0, can we still include it in the plugin release to test on production before syncing to WP core?

Sure, let's include if possible @jeryj comment.

**Comment by richtabor** [comment]
It should always just say "Link", just like the standard link control as seen in the paragraph block.

**Comment by jeryj** [comment]
@richtabor So, the changes needed:

- Button always say "Link"
- Button should be in the active (Black background) state with `aria-pressed="true"` if the "Expand on Click" option is set

**Comment by artemiomorales** [comment]
>- Button always say "Link"
>- Button should be in the active (Black background) state with aria-pressed="true" if the "Expand on Click" option is set

Changes added [here](https://github.com/WordPress/gutenberg/commit/6e1efc6ae50f127c5456876c968d808a132b5d38).

**Comment by cbravobernal** [comment]
This PR has been cherry-picked for `release/17.6` in: https://github.com/WordPress/gutenberg/commit/697495fe7d5bf46b64768012da69bb736e0f7fca

**Comment by t-hamano** [comment]
This change appears to cause an unintended horizontal scrollbar in the MediaReplaceFlow `component`. This is probably due to the width being reduced on [this line](https://github.com/WordPress/gutenberg/pull/57608/files#diff-b6b4e41f983f8b1d918517a1f5b838c88b11b94ad7cb77ce21e1cb7a2ede1b9cR30).

This issue is summarized in #58593.

**Comment by t-hamano** [comment]
Another regression: #58599

**Pull request merged**
